File: p51lee/w02/w02_pgs_dp_exprbyn.ts
function solution(N: number, number: number) {
    let N_string: string = N.toString()
    let series_of_N: [number, Set<number>][] = [1, 2, 3, 4, 5, 6, 7, 8].map(
        (value: number): [number, Set<number>] => [value, new Set([parseInt(N_string.repeat(value))])]
    )
    let solution_map: Map<number, Set<number>> = new Map(series_of_N)

    for (let ord1 = 1; ord1 < 9; ord1++) {
        if (solution_map.get(ord1)!.has(number)) {
            return ord1
        }
        for (let op1 of solution_map.get(ord1)!) {
            for (let ord2 = 1; ord2 <= Math.min(ord1, 8-ord1); ord2++) {
                for (let op2 of solution_map.get(ord2)!) {
                    solution_map.get(ord1+ord2)!
                        .add(op1+op2)
                        .add(Math.abs(op1-op2))
                        .add(op1*op2)
                    if (op2 != 0 && op1 % op2 == 0) {
                        solution_map.get(ord1+ord2)!.add(op1/op2)
                    }
                    if (op1 != 0 && op2 % op1 == 0) {
                        solution_map.get(ord1+ord2)!.add(op2/op1)
                    }
                }
            }
        }
    }
    return -1
}

File: p51lee/w02/w02_pgs_dp_exprbyn.py
def solution(N, number):
    N_string = str(N)
    series_of_N = [int(N_string * i) for i in range(1, 9)]
    solution_dict = {i: {series_of_N[i - 1]} for i in range(1, 9)}

    for ord1 in range(1, 9):
        if number in solution_dict[ord1]:
            return ord1
        for op1 in solution_dict[ord1]:
            for ord2 in range(1, min(ord1, 8 - ord1) + 1):
                for op2 in solution_dict[ord2]:
                    solution_dict[ord1 + ord2].update(
                        [op1 + op2, abs(op1 - op2), op1 * op2]
                    )
                    if op2 != 0 and op1 % op2 == 0:
                        solution_dict[ord1 + ord2].add(op1 / op2)
                    if op1 != 0 and op2 % op1 == 0:
                        solution_dict[ord1 + ord2].add(op2 / op1)

    return -1